### DISCUSSION TITLE
Support for Generic Mapper / ParameterizedType

### DISCUSSION THREAD

**Discussion by amoscatelli** [mapstruct/mapstruct#3261] [Q&A]
Hi there !

Is something like this supported ?
"Mappers" doesn't have a getMapper method with a ParameterizedType argument.

```java
@Mapper
public interface GenericMapper<D, T> {

   D to(T from);
   T from(D to);

}
```

Also it would be nice to be able to inject it 

```java
@Inject
private GenericMapper<UserDTO, User> mapper;
```

This would be usefull, I don't want to have N identical Mapper with the only difference beeing in the DTO/Entity types they convert from/to.

Can I already achieve what I want or should I file a feature request ?

**Comment by filiphr** [comment]
@amoscatelli MapStruct is an annotation processor, which means that it generates code during compile time. Annotating a generic mapper won't do anything, since MapStruct doesn't know what to do with that generic mapper during compile time.

You have to define mappers that extend from your `GenericMapper` and then inject them into your context.

If you do not want to type them by hand then you can write your own annotation processor that is going to generate the MapStruct mappers

**Reply by amoscatelli** [reply]
@filiphr by scanning @Inject annotations MapStruct may be able to know, at compile time, what to do with a generic mapper.

But I suppose this is out of scope ?

Thank you

**Reply by filiphr** [reply]
There might be ways to write an annotation processor that is going to generate what you need. However, this is out of scope for MapStruct.

In theory you can write your own annotation processor that is going to scan `@Inject` and generate a MapStruct interface which in turn will generate the implementation
